cmd/dsl/main.go

```diff
@@ -141,30 +172,36 @@ func (s *Simulator) updateCurrentInstruction(instr Instruction) {
 	s.instr = instr
 }
 
-func listenOnPort(port int) bool {
+func listenOnPort(port int) error {
 	l, err := net.Listen("tcp", fmt.Sprintf("localhost:%d", port))
 	if err != nil {
 		// we couldn't use this port, close the socket & try a new one
-		return false
+		return err
 	}
 	// success! we could listen on the port, which means we can use it!
 	// close the opened socket and return the port
 	l.Close()
-	return true
+	return nil
 }
 
 func (s *Simulator) acquirePort() int {
 	maxAttempts := 100
 	startPort := s.basePort + s.portCounter
+
 	for i := 0; i < maxAttempts; i = i + 2 {
 		port := s.basePort + s.portCounter
 		s.portCounter += 2
 		// try to acquire two sequential ports: one for muxrpc communication, the other for sbot's websockets support.
 		// websockets is currently not used by the netsim, but the port needs to be specified for the sbots process
-		if listenOnPort(port) && listenOnPort(port+1) {
-			// if we could acquire the two ports, we're done! we have found two usable ports for one of our puppets
-			return port
+		g := new(errgroup.Group)
+		g.Go(func() error { return listenOnPort(port) })
```

Comment:
if you want to get even more functional, you could turn `listenOnPort(int) bool` into this:

```go
func listenOnPort(port int) func() error {
	return func() error {
		l, err := net.Listen("tcp", fmt.Sprintf("localhost:%d", port))
		if err != nil {
			// we couldn't use this port, close the socket & try a new one
			return err
		}
		// success! we could listen on the port, which means we can use it!
		// close the opened socket and return the port
		l.Close()
		return nil
	}
}
```

Then this line would become `g.Go(listenOnPort(port))` and `g.Go(listenOnPort(port+1))`

Reply:
ahhhh!!!!!! that makes a lot of sense! i was a bit bugged by the weakboi signature accepted by `errgroup.Go`, but of course you can return a function that conforms to its demands from the goroutine you actually want to run 🧠 🧠